From 684c6c97a1adfc51be55934ce5dbdbcb63e2b0ad Mon Sep 17 00:00:00 2001
From: Infernio <infernio@icloud.com>
Date: Sat, 10 Oct 2020 20:33:27 +0200
Subject: [PATCH 3/5] Centralize readme URL handling

Makes it work on Linux and if the Docs folder is missing. We should
probably still extract the whole thing to a bass.dirs entry, so that we
can save the entire method call.

Does break another balt coupling, which is nice.
---
 Mopy/bash/balt.py                   | 15 +--------------
 Mopy/bash/basher/app_buttons.py     |  9 ++-------
 Mopy/bash/basher/settings_dialog.py | 12 ++++--------
 Mopy/bash/bolt.py                   | 20 ++++++++++++++++++++
 Mopy/bash/patcher/patch_files.py    |  3 +--
 5 files changed, 28 insertions(+), 31 deletions(-)

diff --git a/Mopy/bash/balt.py b/Mopy/bash/balt.py
index 591423ad..6fde22a4 100644
--- a/Mopy/bash/balt.py
+++ b/Mopy/bash/balt.py
@@ -30,7 +30,7 @@
 #--Localization
 #..Handled by bolt, so import that.
 from . import bolt
-from .bolt import GPath, deprint
+from .bolt import GPath, deprint, readme_url
 from .exception import AbstractError, AccessDeniedError, ArgumentError, \
     BoltError, CancelError, SkipError, StateError
 #--Python
@@ -2233,19 +2233,6 @@ def ask_uac_restart(message, title, mopy):
             u'\n\n' + _(u'See the <A href="%(readmePath)s">readme</A> '
                 u'for more information.') % {'readmePath': readme}])[0]
 
-def readme_url(mopy, advanced=False):
-    readme = mopy.join(u'Docs',
-                       u'Wrye Bash Advanced Readme.html' if advanced else
-                       u'Wrye Bash General Readme.html')
-    if readme.exists():
-        readme = u'file:///' + readme.s.replace(u'\\', u'/').replace(u' ',
-                                                                     u'%20')
-    else:
-        # Fallback to Git repository
-        readme = u"http://wrye-bash.github.io/docs/Wrye%20Bash" \
-                 u"%20General%20Readme.html"
-    return readme
-
 class INIListCtrl(wx.ListCtrl):
 
     def __init__(self, parent):
diff --git a/Mopy/bash/basher/app_buttons.py b/Mopy/bash/basher/app_buttons.py
index ec36c179..644b0595 100644
--- a/Mopy/bash/basher/app_buttons.py
+++ b/Mopy/bash/basher/app_buttons.py
@@ -29,7 +29,7 @@
 from .settings_dialog import SettingsDialog
 from .. import bass, bosh, bolt, balt, bush, mod_files, load_order
 from ..balt import ItemLink, Link, Links, SeparatorLink, BoolLink, staticBitmap
-from ..bolt import GPath
+from ..bolt import GPath, readme_url
 from ..env import getJava
 from ..exception import AbstractError
 from ..gui import ClickableImage, EventResult
@@ -675,12 +675,7 @@ class App_Help(StatusBar_Button):
     imageKey, _tip = u'help.%s', _(u"Help File")
 
     def Execute(self):
-        readme_html = bass.dirs[u'mopy'].join(
-            u'Docs', u'Wrye Bash General Readme.html')
-        if readme_html.isfile():
-            webbrowser.open(readme_html.s)
-        else:
-            balt.showError(Link.Frame, _(u'Cannot find General Readme file.'))
+        webbrowser.open(readme_url(mopy=bass.dirs[u'mopy']))
 
 #------------------------------------------------------------------------------
 class App_DocBrowser(StatusBar_Button):
diff --git a/Mopy/bash/basher/settings_dialog.py b/Mopy/bash/basher/settings_dialog.py
index 9fa19534..e37adc08 100644
--- a/Mopy/bash/basher/settings_dialog.py
+++ b/Mopy/bash/basher/settings_dialog.py
@@ -30,7 +30,7 @@
 from .constants import colorInfo, settingDefaults
 from .. import balt, barb, bass, bolt, bosh, bush, env, exception
 from ..balt import colors, Link, Resources
-from ..bolt import deprint, GPath
+from ..bolt import deprint, GPath, readme_url
 from ..gui import ApplyButton, BusyCursor, Button, CancelButton, Color, \
     ColorPicker, DialogWindow, DropDown, HLayout, HorizontalLine, \
     LayoutOptions, OkButton, PanelWin, Stretch, TextArea, TreePanel, VLayout, \
@@ -106,14 +106,10 @@ def _exec_request_restart(self, requesting_setting, restart_params):
         self._requesting_restart.add(requesting_setting)
         self._restart_params.extend(restart_params)
 
-    def _open_readme(self):
+    @staticmethod
+    def _open_readme():
         """Handles a click on the help button by opening the readme."""
-        general_readme = bass.dirs[u'mopy'].join(
-            u'Docs', u'Wrye Bash General Readme.html')
-        if general_readme.isfile():
-            webbrowser.open(general_readme.s)
-        else:
-            balt.showError(self, _(u'Cannot find General Readme file.'))
+        webbrowser.open(readme_url(mopy=bass.dirs[u'mopy']))
 
     def _send_apply(self):
         """Propagates an Apply button click to all child pages."""
diff --git a/Mopy/bash/bolt.py b/Mopy/bash/bolt.py
index 369acec4..b6b24623 100644
--- a/Mopy/bash/bolt.py
+++ b/Mopy/bash/bolt.py
@@ -1879,6 +1879,26 @@ def winNewLines(inString):
     """Converts unix newlines to windows newlines."""
     return reUnixNewLine.sub(u'\r\n',inString)
 
+def readme_url(mopy, advanced=False):
+    """Returns the URL of the WB readme based on the specified Mopy folder."""
+    readme_fname = (u'Wrye Bash Advanced Readme.html' if advanced else
+                    u'Wrye Bash General Readme.html')
+    if os.name == u'nt':
+        # Windows, use Docs in Mopy folder
+        readme = mopy.join(u'Docs', readme_fname)
+    else:
+        # Linux, use global docs
+        readme = GPath(u'/usr/share/doc/wrye-bash').join(readme_fname)
+    if readme.isfile():
+        # Fix backslashes and escape spaces for the URL
+        readme = u'file:///' + readme.s.replace(u'\\', u'/').replace(u' ',
+                                                                     u'%20')
+    else:
+        # Fall back to hosted version
+        readme = u'http://wrye-bash.github.io/docs/' + readme_fname.replace(
+            u' ', u'%20')
+    return readme
+
 # Log/Progress ----------------------------------------------------------------
 #------------------------------------------------------------------------------
 class Log(object):
diff --git a/Mopy/bash/patcher/patch_files.py b/Mopy/bash/patcher/patch_files.py
index a93f3302..a3a49249 100644
--- a/Mopy/bash/patcher/patch_files.py
+++ b/Mopy/bash/patcher/patch_files.py
@@ -28,11 +28,10 @@
 from .. import bush # for game etc
 from .. import bosh # for modInfos
 from .. import bolt # for type hints
-from ..balt import readme_url
 from .. import load_order
 from .. import bass
 from ..brec import MreRecord
-from ..bolt import GPath, SubProgress, deprint, Progress
+from ..bolt import GPath, SubProgress, deprint, Progress, readme_url
 from ..cint import ObModFile, FormID, dump_record, ObCollection, MGEFCode
 from ..exception import BoltError, CancelError, ModError, StateError
 from ..localize import format_date
-- 
2.28.0

